runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -303,14 +309,34 @@ public void processElements(Iterable<WindowedValue<InputT>> values) throws Excep
       windowsToConsider.addAll(processElement(windowToMergeResult, value));
     }
 
-    // Trigger output from any window for which the trigger is ready
-    for (W mergedWindow : windowsToConsider) {
-      ReduceFn<K, InputT, OutputT, W>.Context directContext =
-          contextFactory.base(mergedWindow, StateStyle.DIRECT);
-      ReduceFn<K, InputT, OutputT, W>.Context renamedContext =
-          contextFactory.base(mergedWindow, StateStyle.RENAMED);
-      triggerRunner.prefetchShouldFire(mergedWindow, directContext.state());
-      emitIfAppropriate(directContext, renamedContext);
+    if (!windowsToConsider.isEmpty()) {
```

Comment:
For readability, it would be nice for this to be an early return, too.

But actually all the loops are dead if the condition holds so I'd be inclined leave out the overarching `if` unless you know allocating an iterator is expensive, which it isn't here.

Since these are smallish in-memory loops, it isn't expensive to re-iterate and it could help readability a lot. Consider if you broke each loop out into methods:

 - `void prefetchShouldFire(Collection<W>)`
 - `Collection<W> windowsThatShouldFire(Collection<W>)`
 - `void prefetchEmitPane(Collection<W>)` (here you've merged with above)
 - `void emitPanes(Collection<W>)`.

I'm trying to keep this method under control, as the class is quite difficult already.